In [1]:
import pandas as pd

In [190]:
customers = pd.read_csv("../data/1_raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/1_raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/1_raw/olist_order_items_dataset.csv")
products = pd.read_csv("../data/1_raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/1_raw/olist_sellers_dataset.csv")
payments = pd.read_csv("../data/1_raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/1_raw/olist_order_reviews_dataset.csv")
category_translation = pd.read_csv("../data/1_raw/product_category_name_translation.csv")
geolocation = pd.read_csv("../data/1_raw/olist_geolocation_dataset.csv")

In [191]:
customers = pd.read_csv("../data/1_raw/olist_customers_dataset.csv")


In [192]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [193]:
customers.shape

(99441, 5)

In [194]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 11.0 MB


In [195]:
customers[["customer_city"]]

,customer_city
0,franca
1,sao bernardo do campo
2,sao paulo
3,mogi das cruzes
4,campinas
...,...
99436,sao paulo
99437,taboao da serra
99438,fortaleza
99439,canoas


In [196]:
customers["customer_id"].unique

<bound method Series.unique of 0        06b8999e2fba1a1fbc88172c00ba8bc7
1        18955e83d337fd6b2def6b18a428ac77
2        4e7b3e00288586ebd08712fdd0374a03
3        b2b6027bc5c5109e529d4dc6358b12c3
4        4f2d8ab171c80ec8364f7c12e35b23ad
                       ...               
99436    17ddf5dd5d51696bb3d7c6291687be6f
99437    e7b71a9017aa05c9a7fd292d714858e8
99438    5e28dfe12db7fb50a4b2f691faecea5e
99439    56b18e2166679b8a959d72dd06da27f9
99440    274fa6071e5e17fe303b9748641082c8
Name: customer_id, Length: 99441, dtype: str>

In [197]:
# primary key
customers["customer_id"].is_unique

True

### Primary Key Observation

`customer_id` is unique, so it can be used as the primary key.
`customer_unique_id` is not unique.
The same customer can appear more than once with different `customer_id` values.

# Understanding The Orders Dataset 


In [198]:
orders = pd.read_csv("../data/1_raw/olist_orders_dataset.csv")

In [199]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [200]:
orders.shape

(99441, 8)

In [201]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 21.9 MB


 # all columns are wrong dtype
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  

 Cancelled or undelivered orders may correctly have no delivery date; filling them with an invented date would make your analysis inaccurate.
 

# Primary Key


In [202]:
orders["order_id"].is_unique

True

In [203]:
orders["customer_id"].head()

0    9ef432eb6251297304e76186b10a928d
1    b0830fb4747a6c6d20dea0b8c802d7ef
2    41ce2a54c0b03bf3443c3d931a367089
3    f88197465ea7920adcdbec7375364d82
4    8ab97904e6daea8866dbdbc4fb7aad2c
Name: customer_id, dtype: str

### Verify Customer Relationship

Every `customer_id` in the Orders table exists in the Customers table.

In [204]:
orders["customer_id"].isin(customers["customer_id"]).all()

np.True_

### Orders Table Summary

Primary Key: `order_id`
Foreign Key: `customer_id`
`customer_id` connects the Orders table with the Customers table.
All customer IDs in Orders exist in the Customers table.

# 3.Order Items Dataset


In [205]:
order_items = pd.read_csv("../data/1_raw/olist_order_items_dataset.csv")

In [206]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [207]:
order_items.shape

(112650, 7)

In [208]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 18.4 MB


 #  shipping_limit_date = Change to datetime

# no primary key 

In [209]:
order_items["order_id"].is_unique

False

## Understand The Products Dataset

In [210]:
products = pd.read_csv("../data/1_raw/olist_products_dataset.csv")

In [211]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [212]:
products.shape


(32951, 9)

In [213]:
products.info()


<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 3.7 MB


In [214]:
products.describe()


,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
mean,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000
max,76.000000,3992.000000,20.000000,40425.000000,105.000000,105.000000,118.000000


# product_id is primary key

In [215]:
products["product_id"].is_unique 

True

## Understand The Sellers Dataset

In [216]:
sellers = pd.read_csv("../data/1_raw/olist_sellers_dataset.csv")

In [217]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [218]:
sellers.shape

(3095, 4)

In [219]:
sellers.info()

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 230.3 KB


# Primary key

In [220]:
sellers["seller_id"].is_unique

True

# we can check these relationships:

In [221]:
order_items["product_id"].isin(products["product_id"]).all()

np.True_

In [222]:
order_items["seller_id"].isin(sellers["seller_id"]).all()

np.True_

# Order Items Relationships

`order_id` connects Order Items with Orders.
`product_id` connects Order Items with Products.
`seller_id` connects Order Items with Sellers.
The combination of `order_id` and `order_item_id` acts as a composite key.

## Understand The Payments Dataset

In [223]:
payments = pd.read_csv("../data/1_raw/olist_order_payments_dataset.csv")

In [224]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [225]:
payments.shape

(103886, 5)

In [226]:
payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 8.1 MB


In [227]:
payments["order_id"].is_unique

False

# check the relationship with Orders

In [228]:
payments["order_id"].isin(orders["order_id"]).all()

np.True_

## Understand The Reviews Dataset

In [229]:
reviews = pd.read_csv("../data/1_raw/olist_order_reviews_dataset.csv")

In [230]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [231]:
reviews.shape

(99224, 7)

In [232]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 17.8 MB


 # correct this columns
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str 

In [233]:
reviews["order_id"].is_unique

False

In [234]:
reviews["review_creation_date"].is_unique

False

# check the relationship with Orders

In [235]:
reviews["order_id"].isin(orders["order_id"]).all()

np.True_

### Reviews Table Summary

`review_id` is not unique by itself.
`order_id` connects Reviews with Orders.
All `order_id` values in Reviews exist in the Orders table.


# Understand The Product Category Translation Dataset

In [236]:

category_translation = pd.read_csv( "../data/1_raw/product_category_name_translation.csv")


In [237]:
category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [238]:
category_translation.shape

(71, 2)

In [239]:
category_translation.info()

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 3.5 KB


# Both columns are primary keys in this Table

In [240]:
category_translation["product_category_name"].is_unique

True

In [241]:
category_translation["product_category_name"].unique

<bound method Series.unique of 0                      beleza_saude
1            informatica_acessorios
2                        automotivo
3                   cama_mesa_banho
4                  moveis_decoracao
                  ...              
66                           flores
67               artes_e_artesanato
68                  fraldas_higiene
69    fashion_roupa_infanto_juvenil
70               seguros_e_servicos
Name: product_category_name, Length: 71, dtype: str>

In [242]:
category_translation["product_category_name_english"].is_unique

True

# we can check these relationships with products

In [243]:
category_translation["product_category_name"].isin( products["product_category_name"]).all()

np.True_

# Product Category Translation Summary

`product_category_name` connects with  Products table.



# Understand The Geolocation Dataset

In [244]:
geolocation = pd.read_csv("../data/1_raw/olist_geolocation_dataset.csv")

In [245]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [246]:
geolocation.shape

(1000163, 5)

In [247]:
geolocation.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 50.1 MB


In [248]:
geolocation["geolocation_zip_code_prefix"].is_unique

False

# Geolocation Table Summary

'geolocation_zip_code_prefix` is not unique.

The same ZIP-code prefix can appear multiple times.

This table contains latitude, longitude, city, and state information.

It can help analyze customer and seller locations.

# Data Dictionary and Relationship Summary

# Here are Main Table Relationships

Customers
    ↓ customer_id
Orders
    ↓ order_id
Order Items
    ├── product_id → Products
    └── seller_id → Sellers

Orders
    ├── order_id → Payments
    └── order_id → Reviews

Products
    └── product_category_name → Category Translation

Customers / Sellers
    └── zip_code_prefix → Geolocation


## Data Understanding Conclusion

The 9 datasets have been loaded and examined.

We identified:
- Dataset shapes
- Column names
- Data types
- Primary and composite keys
- Foreign-key relationships

The next step is Data Quality Analysis.
under the below-

# step 4:  Data Quality Analysis

●	Identify missing values

●	Check duplicates

●	Check incorrect data types

●	Identify inconsistent or invalid values

●	Identify potential outliers

●	Check primary-key uniqueness



In [249]:
customers = pd.read_csv("../data/1_raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/1_raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/1_raw/olist_order_items_dataset.csv")
products = pd.read_csv("../data/1_raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/1_raw/olist_sellers_dataset.csv")
payments = pd.read_csv("../data/1_raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/1_raw/olist_order_reviews_dataset.csv")
category_translation = pd.read_csv("../data/1_raw/product_category_name_translation.csv")
geolocation = pd.read_csv("../data/1_raw/olist_geolocation_dataset.csv")

In [250]:
import pandas as pd 

In [251]:
# 1.Check missing values in each column

customers.isnull().sum()


customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [252]:
customers.isnull().mean()*100


customer_id                 0.0
customer_unique_id          0.0
customer_zip_code_prefix    0.0
customer_city               0.0
customer_state              0.0
dtype: float64

In [253]:
#  Count completely duplicated rows
customers.duplicated().sum()



np.int64(0)

In [254]:
# Check column data types
customers.dtypes

customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

# Identify inconsistent or invalid values
This helps you spot different spellings, extra spaces, or unexpected state codes.

In [255]:
customers["customer_state"].value_counts(dropna=False)

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

# 5. Identify potential outliers
- Potential outliers are values unusually high or low compared with most other values.
- Choose a numeric measurement use not any columns
- Even if its data type is integer, you should not use IQR to check outliers.



# Check primary-key uniqueness

Check primary-key uniqueness means checking that the column identifying each row has no repeated values and no missing values.

In [256]:
customers["customer_id"].is_unique

True

In [257]:
customers["customer_state"].is_unique # check the # To display any repeated IDs:

False

# To display any repeated IDs:

In [258]:
customers[
    customers["customer_unique_id"].duplicated(keep=False)
]

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
13,eabebad39a88bb6f5b52376faec28612,295c05e81917928d76245e842748184d,5704,sao paulo,SP
32,2d5831cb2dff7cdefba62e950ae3dc7b,e9dd12dca17352644a959d9dea133935,42800,camacari,BA
33,b2bed119388167a954382cca36c4777f,e079b18794454de9d2be5c12b4392294,27525,resende,RJ
...,...,...,...,...,...
99324,5b46a0d983eec8c97363bea78d4a69dd,8bab3162259edfaadd1ea2e1fe7f58dc,31565,belo horizonte,MG
99327,c1affa46f9f3b514555259049a0307b9,12ab9334b1240d6d037f2b0102a49571,38050,uberaba,MG
99336,ebf46ff530343a129926adc1f831dea4,0ee57f62666561b72f2ceacad0230cbf,9530,sao caetano do sul,SP
99353,282fbce48e4d2077aad602dd125c9225,0ceb502fc33a2ad327b08288c5310e2e,29134,viana,ES


# Check the order Data -2 

In [259]:
# Check missing values in each column
orders.isnull().sum()



order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [260]:
# Count completely duplicated rows
orders.duplicated().sum()



np.int64(0)

In [261]:
# Check column data types
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

# Identify inconsistent or invalid values
This helps you spot different spellings, extra spaces, or unexpected state codes.

In [262]:
orders["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

#  Identify potential outliers
- Use this check for price, freight, or product weight
- This data is not any outlier 

In [263]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 21.9 MB


# primary-key uniqueness check

In [264]:
# Check primary-key uniqueness
orders["order_id"].is_unique

True

In [265]:
orders["customer_id"].is_unique 

True

# 3 Check the Data  order_items 



In [266]:
# Check missing values in each column
order_items.isnull().sum()



order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [267]:
# Count completely duplicated rows
order_items.duplicated().sum()
# reomove duplicate code
# orders[orders["order_id"].duplicated(keep=False)]



np.int64(0)

In [268]:
# Check column data types
order_items.dtypes

order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

# Identify inconsistent or invalid values
This helps you spot different spellings, extra spaces, or unexpected state codes.

- customer_state	 Different spellings, extra spaces, unexpected codes
- price 	Negative amounts
- freight_value 	Negative shipping charges
- review_score	 Scores outside 1–5

# A value of 0 is allowed—it may mean free shipping.

- Check for negative prices, negative freight charges, and item numbers below 1.

- The note “0 may mean free shipping” applies to freight_value, so place it beside the freight check.

In [269]:
order_items[order_items["price"] < 0]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value


In [270]:

order_items[order_items["order_item_id"] < 1]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value


#  Identify potential outliers

I used the IQR method to identify unusually high or low prices and freight charges. It flagged 8,427 rows for price and 12,134 rows for freight value. These records require review before any cleaning decision.

In [271]:
Q1 = order_items["price"].quantile(0.25)
Q3 = order_items["price"].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

order_items[
    (order_items["price"] < lower) |
    (order_items["price"] > upper)
]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
7,000576fe39319847cbb9d288c5617fa6,1,557d850972a7d6f792fd18ae1400d9b6,5996cddab893a4652a15592fb58ab8db,2018-07-10 12:30:45,810.00,70.75
16,0009c9a17f916a706d71784483a5d643,1,3f27ac8e699df3d300ec4a5d8c5cf0b2,fcb5ace8bcc92f75707dc0f01a27d269,2018-05-02 09:31:53,639.00,11.34
26,0011d82c4b53e22e84023405fb467e57,1,c389f712c4b4510bc997cee93e8b1a28,bfd27a966d91cfaafdb25d076585f0da,2018-01-29 21:51:25,289.00,26.33
30,00137e170939bba5a3134e2386413108,1,672e757f331900b9deea127a2a7b79fd,e59aa562b9f8076dd550fcddf0e73491,2017-11-30 06:30:55,397.00,24.65
39,0017afd5076e074a48f1f1a4c7bac9c5,1,fe59a1e006df3ac42bf0ceb876d70969,25c5c91f63607446a97b143d2d535d31,2017-04-24 22:25:19,809.10,44.29
...,...,...,...,...,...,...,...
112586,ffdc5e3279114c523a09296f8fd28331,1,5e70a4508abd5d19bfe7f110785294c8,328acb143b4a73aad6bee8e734f5d20d,2017-02-14 14:49:23,299.99,17.80
112592,ffde92ba447b33a47d1c04d203f10f41,1,a97cd00f5c1dbfbc7f0f4b6c1032253e,b347677812ea483b0f528eaf8cbc09b7,2017-11-30 14:13:18,278.00,26.70
112632,fff7c4452f050315db1b3f24d9df5fcd,1,dd469c03ad67e201bc2179ef077dcd48,7e93a43ef30c4f03f38b393420bc753a,2017-06-07 17:05:23,736.00,20.91
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41


In [272]:
Q1 = order_items["freight_value"].quantile(0.25)
Q3 = order_items["freight_value"].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

order_items[
    (order_items["freight_value"] < lower) |
    (order_items["freight_value"] > upper)
]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
7,000576fe39319847cbb9d288c5617fa6,1,557d850972a7d6f792fd18ae1400d9b6,5996cddab893a4652a15592fb58ab8db,2018-07-10 12:30:45,810.00,70.75
22,000f25f4d72195062c040b12dce9a18a,1,1c05e0964302b6cf68ca0d15f326c6ba,7c67e1448b00f6e969d365cea6b010ab,2018-03-21 11:10:11,119.99,44.40
39,0017afd5076e074a48f1f1a4c7bac9c5,1,fe59a1e006df3ac42bf0ceb876d70969,25c5c91f63607446a97b143d2d535d31,2017-04-24 22:25:19,809.10,44.29
59,00254baeb6c932b0a8aeead91fbd02b5,1,18901878788fec7ddc55e64d1ace8187,e8b4225284fbb02d16f200513f1f395d,2018-05-14 22:14:46,149.90,43.11
73,002b430ff89b3a24c31a1170acbbedea,1,cc47c0863559499f01b4cad750929d10,7a67c85e85bb2ce8582c35f2203ad736,2017-06-26 22:10:14,199.99,65.56
...,...,...,...,...,...,...,...
112641,fffb9224b6fc7c43ebb0904318b10b5f,2,43423cdffde7fda63d0414ed38c11a73,b1fc4f64df5a0e8b6913ab38803c57a9,2017-11-03 02:55:58,55.00,34.19
112642,fffb9224b6fc7c43ebb0904318b10b5f,3,43423cdffde7fda63d0414ed38c11a73,b1fc4f64df5a0e8b6913ab38803c57a9,2017-11-03 02:55:58,55.00,34.19
112643,fffb9224b6fc7c43ebb0904318b10b5f,4,43423cdffde7fda63d0414ed38c11a73,b1fc4f64df5a0e8b6913ab38803c57a9,2017-11-03 02:55:58,55.00,34.19
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41


# primary-key uniqueness check


In [273]:
# Check repeated order_id + order_item_id pairs
order_items.duplicated(
    subset=["order_id", "order_item_id"]
).sum()

np.int64(0)

In [274]:
order_items["order_id"].is_unique

False

# Check The 4. Product Data

In [275]:
# Check missing values in each column
products.isnull().sum()



product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [276]:
# Count completely duplicated rows
products.duplicated().sum()



np.int64(0)

In [277]:
# Check column data types
products.dtypes

product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [278]:
products

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0
...,...,...,...,...,...,...,...,...,...
32946,a0b7d5a992ccda646f2d34e418fff5a0,moveis_decoracao,45.0,67.0,2.0,12300.0,40.0,40.0,40.0
32947,bf4538d88321d0fd4412a93c974510e6,construcao_ferramentas_iluminacao,41.0,971.0,1.0,1700.0,16.0,19.0,16.0
32948,9a7c6041fa9592d9d9ef6cfe62a71f8c,cama_mesa_banho,50.0,799.0,1.0,1400.0,27.0,7.0,27.0
32949,83808703fc0706a22e264b9d75f04a2e,informatica_acessorios,60.0,156.0,2.0,700.0,31.0,13.0,20.0


# Identify inconsistent or invalid values
This helps you spot different spellings, extra spaces, or unexpected state codes.

- customer_state	 Different spellings, extra spaces, unexpected codes
- price 	Negative amounts
- freight_value 	Negative shipping charges
- review_score	 Scores outside 1–5

In [279]:
products["product_category_name"].value_counts(dropna=False)

product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 74, dtype: int64

# 5. Potential outliers in product weight

In [280]:
Q1 = products["product_weight_g"].quantile(0.25)
Q3 = products["product_weight_g"].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

products[
    (products["product_weight_g"] < lower) |
    (products["product_weight_g"] > upper)
]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
6,732bd381ad09e530fe0a5f457d81becb,cool_stuff,56.0,1272.0,4.0,18350.0,70.0,24.0,44.0
11,03b63c5fc16691530586ae020c345514,bebes,49.0,728.0,4.0,7150.0,50.0,19.0,45.0
33,1eba879220bd0981a0e2fbab499ed4e1,moveis_escritorio,57.0,476.0,1.0,8950.0,52.0,51.0,17.0
35,5370b82a213393979691c98074265584,bebes,52.0,708.0,9.0,7950.0,33.0,30.0,44.0
36,b8dac5113b06a97e64943234522572b9,automotivo,43.0,1062.0,2.0,6000.0,87.0,3.0,42.0
...,...,...,...,...,...,...,...,...,...
32896,180fbb426ea5ab0708d43ad3f3839bb9,moveis_sala,45.0,843.0,3.0,7600.0,61.0,11.0,46.0
32929,01cb96c64e8e35f29826a58963f6ecc5,construcao_ferramentas_iluminacao,39.0,990.0,2.0,7100.0,43.0,23.0,43.0
32941,6ec96c91757fad0aecafc0ee7f262dcc,bebes,62.0,1417.0,1.0,9550.0,36.0,35.0,35.0
32943,3becff10d1deb92b02f2a1ee62a04524,informatica_acessorios,54.0,1520.0,2.0,6150.0,30.0,30.0,20.0


# 6. Primary-key uniqueness

In [281]:
products["product_id"].is_unique

True

# 5.Data seller

In [282]:
# Check missing values in each column
sellers.isnull().sum()



seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [283]:
# Count completely duplicated rows
sellers.duplicated().sum()



np.int64(0)

In [284]:
# Check column data types
sellers.dtypes

seller_id                   str
seller_zip_code_prefix    int64
seller_city                 str
seller_state                str
dtype: object

# Identify inconsistent or invalid values

In [285]:
sellers["seller_state"].value_counts(dropna=False)

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
AC       1
PI       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

#  Identify potential outliers


In [286]:
# Not applicable: this table contains IDs and location details.
# ZIP prefixes are codes, so do not apply numeric outlier checks.

# primary-key uniqueness check

In [287]:
sellers["seller_id"].is_unique

True

# 6.Check the payments Data 

In [288]:
# Check missing values in each column
payments.isna().sum()



order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [289]:
# Count completely duplicated rows
payments.duplicated().sum()



np.int64(0)

In [290]:
# Check column data types
payments.dtypes

order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

# Identify inconsistent or invalid values

In [291]:
payments["payment_type"].value_counts(dropna=False)

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [292]:
payments[
    (payments["payment_value"] < 0) |
    (payments["payment_installments"] < 0) |
    (payments["payment_sequential"] < 1)
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value


In [293]:
payments[payments["payment_value"] == 0]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


#  Identify potential outliers


In [294]:
Q1 = payments["payment_value"].quantile(0.25)
Q3 = payments["payment_value"].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

payments[
    (payments["payment_value"] < lower) |
    (payments["payment_value"] > upper)
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
23,4d680edbaa7d3d9bed69532957368a03,1,credit_card,10,353.09
26,d0a945f85ba1074b60aac97ade7e240e,1,credit_card,2,541.00
52,2e2c60b99754ae1e4d8b18846cfec9f2,1,credit_card,4,542.66
54,95442deb81a5d91c97c0df96b431634a,1,boleto,1,368.98
61,3ab4eea6776259ae8dfbabaeffece352,1,credit_card,8,387.80
...,...,...,...,...,...
103814,df8ab61cb615d551483618533aceb364,1,credit_card,10,713.65
103843,262bf694d9c87de7844dd80b98fcfef3,1,credit_card,5,372.08
103861,691688405bcee3ff9023639514498962,1,credit_card,1,356.53
103881,0406037ad97740d563a178ecc7a2075c,1,boleto,1,363.31


# primary-key uniqueness check

In [295]:
payments.duplicated(
    subset=["order_id", "payment_sequential"]
).sum()

np.int64(0)

# 7. Check the Review data 

In [296]:
# Check missing values in each column
reviews.isna().sum()



review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [297]:
# Count completely duplicated rows
reviews.duplicated().sum()



np.int64(0)

In [298]:
# Check column data types
reviews.dtypes

review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date         str
review_answer_timestamp      str
dtype: object

# Identify inconsistent or invalid values

In [299]:
reviews[~reviews["review_score"].isin([1, 2, 3, 4, 5])]

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp


#  Identify potential outliers


In [300]:
# Review scores have a fixed range of 1–5.
# Use the valid-range check above; IQR is not needed.

# primary-key uniqueness check

In [301]:
reviews["review_id"].is_unique

False

# 8. Check the category_translation Data

In [302]:
# Check missing values in each column
category_translation.isnull().sum()


product_category_name            0
product_category_name_english    0
dtype: int64

In [303]:

# Count completely duplicated rows
category_translation.duplicated().sum()



np.int64(0)

In [304]:
# Check column data types
category_translation.dtypes

product_category_name            str
product_category_name_english    str
dtype: object

# Identify inconsistent or invalid values

In [305]:
category_translation[
    (category_translation["product_category_name"].str.strip() == "") |
    (category_translation["product_category_name_english"].str.strip() == "")
]

,product_category_name,product_category_name_english


#  Identify potential outliers


In [306]:
# Not applicable: both columns contain category names.
# There are no numeric measurements for an IQR check.

# primary-key uniqueness check

In [307]:
category_translation["product_category_name"].is_unique

True

# 8. Check the geolocation Data

In [308]:
# Check missing values in each column
geolocation.isnull().sum()



geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [309]:
# Count completely duplicated rows
geolocation.duplicated().sum()



np.int64(261831)

In [310]:
# Check column data types
geolocation.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                   str
geolocation_state                  str
dtype: object

# Identify inconsistent or invalid values

In [311]:
geolocation["geolocation_state"].value_counts(dropna=False)

geolocation_state
SP    404268
MG    126336
RJ    121169
RS     61851
PR     57859
SC     38328
BA     36045
GO     20139
ES     16748
PE     16432
DF     12986
MT     12031
CE     11674
PA     10853
MS     10431
MA      7853
PB      5538
RN      5041
PI      4549
AL      4183
TO      3576
SE      3563
RO      3478
AM      2432
AC      1301
AP       853
RR       646
Name: count, dtype: int64

# Identify potential outliers

# Check primary-key uniqueness

In [312]:
geolocation["geolocation_zip_code_prefix"].is_unique

False

# Step 5: Data Cleaning & Preprocessing

●	Handle missing values

●	Remove duplicates

●	Correct data types

●	Format date/time columns

●	Standardize categorical values

●	Handle invalid records

●	Rename columns where required


In [313]:
customers = pd.read_csv("../data/1_raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/1_raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/1_raw/olist_order_items_dataset.csv")
products = pd.read_csv("../data/1_raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/1_raw/olist_sellers_dataset.csv")
payments = pd.read_csv("../data/1_raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/1_raw/olist_order_reviews_dataset.csv")
category_translation = pd.read_csv("../data/1_raw/product_category_name_translation.csv")
geolocation = pd.read_csv("../data/1_raw/olist_geolocation_dataset.csv")

# step 1:- 1. Handle missing values
1. Customers Dataset

- Missing values	 0 in every column	 No filling needed

- Duplicate rows	0	No removal needed

- Duplicate customer_id	0	Suitable primary key

- Extra spaces around text	0	No trimming needed

- City formatting	All lowercase	Already consistent

- State codes	All valid abbreviations	No correction needed

- ZIP prefixes	All five-digit strings	Keep as text


In [314]:
import pandas as pd

customers = pd.read_csv(
    "../Data/1_raw/olist_customers_dataset.csv",
    dtype={"customer_zip_code_prefix": "string"}
)

customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,09790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,01151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,08775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [315]:
customers.to_csv("../Data/2_cleaned/customers_clean.csv", index=False)

In [316]:
#.str.strip() removes spaces at the beginning and end.
#.str.lower() makes city names lowercase.
#.str.upper() makes state codes uppercase
# We use this code to keep city names and state codes in a consistent format.

customers["customer_city"] = customers["customer_city"].str.strip().str.lower()

customers["customer_state"] = customers["customer_state"].str.strip().str.upper()

customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,09790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,01151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,08775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


## Orders Data Cleaning

In [317]:
orders = pd.read_csv("../Data/1_raw/olist_orders_dataset.csv")

orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [318]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [319]:
orders[orders["order_approved_at"].isnull()]["order_status"].value_counts()

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

In [320]:
orders[orders["order_delivered_carrier_date"].isnull()]["order_status"].value_counts()

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

In [321]:
missing_delivery = orders[orders["order_delivered_customer_date"].isnull()]

missing_delivery["order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

Retain missing dates because actual dates are unknown.

Some missing dates relate to orders that have not completed delivery.

Delivered orders with missing dates are data quality issues.

In [322]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Missing dates are retained because the actual dates are unknown.

No dates are filled, and no rows are deleted.

# Remove duplicates


order data is not any duplicate rows

In [323]:
orders.duplicated().sum()

np.int64(0)

# Correct data types


In [324]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

# Format date/time columns


In [325]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])

orders["order_approved_at"] = pd.to_datetime(orders["order_approved_at"])

orders["order_delivered_carrier_date"] = pd.to_datetime(orders["order_delivered_carrier_date"])

orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])

orders["order_estimated_delivery_date"] = pd.to_datetime(orders["order_estimated_delivery_date"])

In [326]:
orders.dtypes

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

# Standardize categorical values


In [327]:
orders["order_status"] = orders["order_status"].str.strip().str.lower()

orders["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

# Handle invalid records

any order was delivered before it was purchased.

A package should be handed to the delivery company first, then received by the customer.


In [328]:
#The code finds records where customer delivery happened before carrier handover, then counts them.

invalid_delivery = orders[
    orders["order_delivered_customer_date"] < orders["order_purchase_timestamp"]
]

invalid_delivery.shape[0]

0

This only checks the records. It does not delete or change them.

In [329]:
invalid_shipping = orders[
    orders["order_delivered_customer_date"] <
    orders["order_delivered_carrier_date"]
]

invalid_shipping.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
6437,a1abeb653a4d4cd1e142ccb8c82cd069,5f50465da00b7fed5dd1239f4ecf6e2c,delivered,2017-07-20 11:20:52,2017-07-21 06:43:14,2017-07-28 16:57:58,2017-07-25 19:32:56,2017-08-14
9553,383aa8b2724fe452d9ccd9934a8c628b,b1cb2f9d7a19480f3749e248db14d58f,delivered,2017-07-02 20:58:43,2017-07-02 21:10:20,2017-07-07 17:22:41,2017-07-06 14:27:51,2017-07-21
13487,cb1134f9010d242e9515ad1c78ec0c39,2fd33ac77677bd214b1882868317eeed,delivered,2017-07-16 12:35:34,2017-07-18 06:03:50,2017-07-20 19:22:02,2017-07-19 14:13:28,2017-08-08
14474,dceb62e8fa94b46006c9554fed743df0,2721900eb4e0f1cc2c836dd7bc1b1e11,delivered,2017-07-20 20:58:05,2017-07-22 11:45:11,2017-08-01 18:23:30,2017-07-26 18:09:10,2017-08-11
19268,5f9d46795c3126674e52becb3a1a517f,79287bcaafdde5c793b996fc40bb7d9f,delivered,2017-07-18 11:48:20,2017-07-18 12:03:29,2017-07-20 23:03:42,2017-07-20 18:52:41,2017-07-31


In [330]:
invalid_delivery = orders[
    orders["order_delivered_customer_date"] < orders["order_purchase_timestamp"]
]

invalid_delivery.shape[0]

0

- If the count shows 0, no orders were delivered to the customer before being handed to the carrier. No correction is needed for this check.


- save your cleaned orders

In [ ]:
#- save your cleaned orders 
orders.to_csv("../Data/2_cleaned/orders_clean.csv", index=False)

# Order Items — Data Cleaning & Preprocessing

In [332]:
order_items = pd.read_csv("../Data/1_raw/olist_order_items_dataset.csv")

order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


# check for missing values
This  column has no missing values.

In [333]:
# Check missing values in each column
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

# check for duplicate rows:

In [334]:
# Count fully duplicated rows
order_items.duplicated().sum()

np.int64(0)

# check the data types.

In [335]:
# Check data types
order_items.dtypes

order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

In [ ]:
# Convert shipping deadline from text to datetime
order_items["shipping_limit_date"] = pd.to_datetime(order_items["shipping_limit_date"]
)

# Verify the data types
order_items.dtypes

order_id                          str
order_item_id                   int64
product_id                        str
seller_id                         str
shipping_limit_date    datetime64[us]
price                         float64
freight_value                 float64
dtype: object

check for negative prices or shipping charges.

In [ ]:
# Count negative values
(order_items[["price", "freight_value"]] < 0).sum()

#One order can contain item 1, item 2, and item 3, but each order_id + order_item_id combination should be unique.

price            0
freight_value    0
dtype: int64

In [338]:
# Save cleaned order items
order_items.to_csv("../Data/2_cleaned/order_items_clean.csv", index=False)

# 4. clean the products dataset
- Products Data

In [388]:
# Load products data
products = pd.read_csv("../Data/1_raw/olist_products_dataset.csv")

# Check missing values
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [389]:
# Label missing product categories
products["product_category_name"] = products["product_category_name"].fillna("unknown")

# Verify remaining missing categories
products["product_category_name"].isnull().sum()

np.int64(0)

In [390]:
products.isnull().sum()

product_id                      0
product_category_name           0
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [391]:
products["product_name_lenght"] = products["product_name_lenght"].fillna("unknown")

# Verify remaining missing categories
products["product_name_lenght"].isnull().sum()

np.int64(0)

In [392]:
products["product_description_lenght"] = products["product_description_lenght"].fillna("unknown")

# Verify remaining missing categories
products["product_description_lenght"].isnull().sum()

np.int64(0)

In [393]:
products["product_weight_g"] = products["product_weight_g"].fillna("unknown")

# Verify remaining missing categories
products["product_photos_qty"].isnull().sum()

np.int64(610)

In [394]:
products["product_photos_qty"] = products["product_photos_qty"].fillna("unknown")

# Verify remaining missing categories
products["product_photos_qty"].isnull().sum()

np.int64(0)

In [395]:
products["product_weight_g"] = products["product_weight_g"].fillna("unknown")

# Verify remaining missing categories
products["product_weight_g"].isnull().sum()

np.int64(0)

In [396]:
products["product_weight_g"] = products["product_weight_g"].fillna("unknown")

# Verify remaining missing categories
products["product_weight_g"].isnull().sum()

np.int64(0)

In [397]:
products["product_length_cm"] = products["product_length_cm"].fillna("unknown")

# Verify remaining missing categories
products["product_length_cm"].isnull().sum()

np.int64(0)

In [398]:
products["product_height_cm"] = products["product_height_cm"].fillna("unknown")

# Verify remaining missing categories
products["product_height_cm"].isnull().sum()

np.int64(0)

In [399]:
products["product_width_cm"] = products["product_width_cm"].fillna("unknown")

# Verify remaining missing categories
products["product_width_cm"].isnull().sum()

np.int64(0)

In [400]:
products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

Missing product categories are labelled "unknown".

Missing numeric details are retained because the actual values are unknown.

# Check duplicated

In [401]:
# Count duplicate product rows
products.duplicated().sum()

np.int64(0)

# check the data types:

In [402]:
products.dtypes

product_id                       str
product_category_name            str
product_name_lenght           object
product_description_lenght    object
product_photos_qty            object
product_weight_g              object
product_length_cm             object
product_height_cm             object
product_width_cm              object
dtype: object

# 5.standardize the category names:

In [403]:
# Remove outer spaces and use lowercase
products["product_category_name"] = (
    products["product_category_name"].str.strip().str.lower()
)

In [404]:
# Count repeated product IDs
products["product_id"].duplicated().sum()

np.int64(0)

cleaned products dataset:

In [ ]:
# Save cleaned products data
products.to_csv("../Data/2_cleaned/products_clean.csv", index=False)


# 5. clean the Sellers dataset
- Sellers Data

In [376]:
# Load sellers data
sellers = pd.read_csv(
    "../Data/1_raw/olist_sellers_dataset.csv",
    dtype={"seller_zip_code_prefix": "string"}
)

# Check missing values
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

# Check duplicated

In [377]:
sellers.duplicated().sum()

np.int64(0)

# check the data types:

In [378]:
sellers.dtypes

seller_id                    str
seller_zip_code_prefix    string
seller_city                  str
seller_state                 str
dtype: object

In [381]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,04195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [385]:
sellers["seller_zip_code_prefix"]= sellers["seller_zip_code_prefix"].astype(int)


In [386]:
sellers.dtypes

seller_id                   str
seller_zip_code_prefix    int64
seller_city                 str
seller_state                str
dtype: object

In [ ]:
# Save cleaned products data
products.to_csv("../Data/1_raw/olist_sellers_dataset.csv", index=False)

# 6. Clean the Payments Data


In [408]:
# Load payments data
payments = pd.read_csv(
        "../data/1_raw/olist_order_payments_dataset.csv",
    dtype={"seller_zip_code_prefix": "string"}
)

# Check missing values
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [409]:
payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 8.1 MB


In [410]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


# Check duplicated

In [ ]:
payments.duplicated().sum()

np.int64(0)

# step 3: Check the data type

In [412]:
payments.dtypes


order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

# step 4: Standardize categorical values



- Removed extra spaces from `payment_type`.
- Converted payment types to lowercase.
- Replaced `not_defined` with `unknown`.

In [418]:
payments["payment_type"] = ( payments["payment_type"].str.strip().str.lower())

# step 5: Rename columns where required

In [420]:
payments["payment_type"] = (
    payments["payment_type"]
    .replace("not_defined", "unknown")
)

In [422]:
payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
unknown            3
Name: count, dtype: int64

# Payments Cleaning Summary

- No duplicate rows were found.
- `payment_type` values were standardized.
- Extra spaces were removed.
- Payment types were converted to lowercase.
- `not_defined` values were replaced with `unknown`.

In [426]:
# Save cleaned payments data
payments.to_csv("../data/1_raw/olist_order_payments_dataset.csv", index=False)

# Clean 7: Reviews dataset cleaning.

In [ ]:
reviews = pd.read_csv("../data/1_raw/olist_order_reviews_dataset.csv")


In [430]:
# Load payments data
reviews = pd.read_csv(
        "../data/1_raw/olist_order_reviews_dataset.csv",
    dtype={"seller_zip_code_prefix": "string"}
)

# Check missing values
reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [440]:
reviews["review_comment_title"] = (
    reviews["review_comment_title"].fillna("No Title")
)

reviews["review_comment_message"] = (
    reviews["review_comment_message"].fillna("No Comment")
)

In [441]:
reviews.isnull().sum()

review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

# 2.check duplicated

In [ ]:
reviews.duplicated().sum()

np.int64(0)

# step 3: Check the data type

No data type correction is required.

In [433]:
reviews.dtypes

review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date         str
review_answer_timestamp      str
dtype: object

# step 4:Format date/time columns

In [436]:
reviews["review_creation_date"] = pd.to_datetime(
    reviews["review_creation_date"],
    errors="coerce"
)

reviews["review_answer_timestamp"] = pd.to_datetime(
    reviews["review_answer_timestamp"],
    errors="coerce"
)

In [437]:
reviews.dtypes

review_id                             str
order_id                              str
review_score                        int64
review_comment_title                  str
review_comment_message                str
review_creation_date       datetime64[us]
review_answer_timestamp    datetime64[us]
dtype: object

# step 5: Standardize categorical values

In [432]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


# Step 6: Check review scores

In [438]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

### Reviews Cleaning Summary

- Checked missing values.
- Missing review titles were replaced with `No Title`.
- Missing review messages were replaced with `No Comment`.
- Checked duplicate rows.
- Converted review date columns to datetime.
- Checked that review scores are between 1 and 5.
- Removed extra spaces from review text.

In [460]:
reviews.to_csv(
    "../Data/2_cleaned/reviews_clean.csv",
    index=False,
    encoding="utf-8"
)

In [443]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,No Title,No Comment,2018-01-18,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,No Title,No Comment,2018-03-10,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,No Title,No Comment,2018-02-17,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,No Title,Recebi bem antes do prazo estipulado.,2017-04-21,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,No Title,Parabéns lojas lannister adorei comprar pela I...,2018-03-01,2018-03-02 10:26:53


# Clean 7: geolocation dataset cleaning.

In [444]:
# Load payments data
geolocation = pd.read_csv(
        "../data/1_raw/olist_geolocation_dataset.csv",
    dtype={"geolocation_zip_code_prefix": "string"}
)

# Check missing values
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [445]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,01037,-23.545621,-46.639292,sao paulo,SP
1,01046,-23.546081,-46.644820,sao paulo,SP
2,01046,-23.546129,-46.642951,sao paulo,SP
3,01041,-23.544392,-46.639499,sao paulo,SP
4,01035,-23.541578,-46.641607,sao paulo,SP


# step  2: dupicated values

In [446]:
geolocation.duplicated().sum()

np.int64(261831)

In [448]:
geolocation.drop_duplicates(inplace=True)

In [452]:
geolocation.duplicated().sum()

np.int64(0)

### Duplicate Cleaning


- 261,831 duplicate rows were found in the Geolocation dataset.
- Duplicate rows were removed from `geolocation_clean`.
- The original `geolocation` dataset was kept unchanged.

# Step 3: Check data types

In [453]:
geolocation.dtypes

geolocation_zip_code_prefix     string
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                   str
geolocation_state                  str
dtype: object

### Geolocation Data Type Check

- ZIP code prefix is stored as integer.
- Latitude and longitude are stored as float.
- City and state are stored as string.
- No data type correction is required.

# Step 4: Standardize categorical values for geolocation_city and geolocation_state.

In [456]:
geolocation["geolocation_city"] = (
    geolocation["geolocation_city"]
    .str.strip()
    .str.lower()
)

geolocation["geolocation_state"] = (
    geolocation["geolocation_state"]
    .str.strip()
    .str.upper()
)

In [457]:
geolocation_clean[["geolocation_city", "geolocation_state"]].head()

,geolocation_city,geolocation_state
0,sao paulo,SP
1,sao paulo,SP
2,sao paulo,SP
3,sao paulo,SP
4,sao paulo,SP


### Categorical Value Standardization

- Removed extra spaces from city and state values.
- Converted city names to lowercase.
- Converted state codes to uppercase.

In [458]:
geolocation_clean.to_csv(
    "../Data/2_cleaned/geolocation_clean.csv",
    index=False,
    encoding="utf-8"
)